src/Aquifer.Jobs/TrackResourceContentRequestJob.cs

```diff
@@ -62,32 +63,27 @@ private async Task LookupIpAddressAsync(TrackResourceContentRequestMessage track
                 return;
             }
 
-            var ipDataRecord = await dbContext.IpAddressData
-                .AsTracking()
-                .SingleOrDefaultAsync(x => x.IpAddress == trackingMetadata.IpAddress, ct);
-            if (ipDataRecord is null)
+            using (var transaction = await dbContext.Database.BeginTransactionAsync(IsolationLevel.Serializable, ct))
```

Comment:
By default selects are not included in the EF generated transaction created via `SaveChangesAsync()`.  By making the transaction here explicit the read is now included in the same transaction as the write.  The `Serializable` isolation level will cause SQL Server to lock relevant rows and to prevent other transaction from concurrently accessing those rows.  Thus, the duplicate key exception should be prevented.

Reply:
The problem, as I understand it, is two situations where `ipDataRecord` is null, so it tries to do two separate inserts with the same PK (the IP Address). How does the select play into that?

Reply:
The DB will lock any rows (and prevent any potential new rows created by other transactions) for our read query by using a range lock, even if there are no results.  By using a `Serializable` transaction the DB guarantees that there will still be no results for that same read query when we commit the writes later.  If we use a different transaction isolation level like `ReadUncommitted` then you wouldn't have the same guarantee about reads (and it wouldn't do range locking).

Note that range locking is more expensive and could result in a full table lock for the duration of the transaction.  If this change causes transaction failures due to that locking then we could switch it back to (a) catch an exception and (b) swallow the error with no logging, but only if (c) the exception is specifically a SQL duplicate key error (by looking at the error number).